include/mbedtls/ssl.h

```diff
@@ -819,19 +841,9 @@ struct mbedtls_ssl_context
     /*
      * Record layer (incoming data)
      */
-    unsigned char *in_buf;      /*!< input buffer                     */
-    unsigned char *in_ctr;      /*!< 64-bit incoming message counter
-                                     TLS: maintained by us
-                                     DTLS: read from peer             */
-    unsigned char *in_hdr;      /*!< start of record header           */
-    unsigned char *in_len;      /*!< two-bytes message length field   */
-    unsigned char *in_iv;       /*!< ivlen-byte IV                    */
-    unsigned char *in_msg;      /*!< message contents (in_iv+ivlen)   */
-    unsigned char *in_offt;     /*!< read offset in application data  */
-
-    int in_msgtype;             /*!< record header: message type      */
-    size_t in_msglen;           /*!< record header: message length    */
-    size_t in_left;             /*!< amount of data read so far       */
+
+    mbedtls_ssl_record in;
```

Comment:
This is an API change, since the ssl context is public

Reply:
Yes, I noted this changes the public API on the original writeup at the top of the PR.... also applies to your comment below.  I think it's quite expensive to duplicate the in and out code with special members for each.

The trend for a long while in OpenSSL has been to make more and more internals opaque, adding accessors where necessary.  They now have few structs elaborated in the public api.  That allows things like, eg, member reorder for struct packing optimization that is needed for the ssl context currently, but is deferred because "it will change the public api".